platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,277 @@
+Event Intake module).
+
+#### Self Events
+
+Events are not only given to the event intake system through gossip. Internal events are also fed to the event intake
+system. These internal events **may** bypass some steps in the pipeline. For example, internal self-events (those
+events created by the node itself) do not need validation. Likewise, when replaying events from the pre-consensus
+recording system, those checks are not needed (since they have already been proved valid and are in topological order).
+
+#### Peer Discipline
+
+During the validation process, if an event is invalid, it is rejected, and this information is passed to the Bad Node
+module so the offending node may be disciplined. Note that the node to be disciplined will be the node that sent this
+bad event to us, not the origin node. This information (which node sent the event) must be captured by Gossip and
+passed to Event Intake as part of the event metadata.
+
+#### Topological Ordering
+
+Events are buffered if necessary to ensure that each parent event has been emitted from the Event Intake before any
+child events. A simple map (the same used for deduplication) can be used here. Given some event, for each parent, look
+up the parent by its hash. If each parent is found in the map, then emit the event. Otherwise, remember the event so
+when the missing parent is received, the child may be emitted. The current implementation uses what is known as the
+"orphan buffer" for this purpose.
+
+Since Event Intake will also maintain some buffers, it needs to know about the progression of the hashgraph,
+so it can evict old events. In this case, the "orphan buffer" holds events until either the parent events have
+arrived, or the events expired due to the advancement of the "non-ancient event window" and the event is
+dropped from the buffer. This document does not prescribe the existence of the orphan buffer or the method by which
+events are sorted and emitted in topological order, but it does describe a method by which old events can be dropped.
+
+#### Emitting Events
+
+When the Event Intake module emits valid, topologically sorted events, it sends them to:
+- The Event Creator module, so it may have information on which events are available to be built on top of as
+  "other parents"
+- The Execution layer as a "pre-handle" event
+- The Pre-consensus Recording module, so the event can be made durable prior to being added to the hashgraph or gossiped
+
+The call to each of these systems is "fire and forget". Specifically, there is no guarantee to Execution that it will
+definitely see an event via `pre-handle` prior to seeing it in `handle`. Technically, Consensus always calls
+`pre-handle` first, but that thread may be parked arbitrarily long by the system and the `handle` thread may actually
+execute first. This is extremely unlikely, but must be defended against in the Execution layer.
```

Comment:
> This is extremely unlikely, but must be defended against in the Execution layer.

This seems like unnecessary complexity to me. If we make the API calls to send events from intake sequentially, only forwarding an event to the `ConsensusEngine` after having notified "pre-handle", then the app can safely operate under the assumption that *every* consensus event was received in "pre-handle"

This "no guarantee of pre-handle" seems like a relic of the current poor design, rather than an unavoidable property of the system.

Reply:
We discussed this, and decided to leave the caveat stated, so that no design is created that relies on strict sequential behavior